### ISSUE TITLE
App start fails with error removing TMPDIR [Dokku 0.5.2] 

### ISSUE THREAD

**Issue by seantanly** [dokku/dokku#2040]
Description of problem:
After upgrading to Dokku version 0.5.2, I'm encountering the following problem which causes the container startup being rolled back. Essentially it seems that in `cleanup()`, rm is unable to remove the `TMPDIR` created.

Running `dokku ps:restart <app>`, everything seems fine until the final bit to cleanup.
```
=====> end <app> web container output
+ cleanup
+ declare 'desc=cleans up TMPDIR and print container output'
+ local id=
+ rm -rf /tmp/user/1000
rm: cannot remove ‘/tmp/user/1000’: Permission denied
+ kill_new bf7b0ffea6034f07a397131f31aebe415a0910e8662c117c4563df8a2e7f6962
+ declare 'desc=wrapper function to kill newly started app container'
+ local id=bf7b0ffea6034f07a397131f31aebe415a0910e8662c117c4563df8a2e7f6962
+ docker inspect bf7b0ffea6034f07a397131f31aebe415a0910e8662c117c4563df8a2e7f6962
+ docker stop bf7b0ffea6034f07a397131f31aebe415a0910e8662c117c4563df8a2e7f6962
+ docker kill bf7b0ffea6034f07a397131f31aebe415a0910e8662c117c4563df8a2e7f6962
```

I have tried rebuilding a fresh dokku instance from scratch and deploying my app, but the deployment fails with the same problem. The app runs fine on Dokku 0.4.14.

```
$ ll /tmp
drwx--x--x  3 root root 4096 Mar 29 08:21 user/

$ ll /tmp/user
drwx------ 2 dokku root 4096 Mar 29 08:34 1000/

$ uname -a
Linux App0 3.13.0-71-generic #114-Ubuntu SMP Tue Dec 1 02:34:22 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

$ docker version
Client:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.3
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   20f81dd
 Built:        Thu Mar 10 15:54:52 2016
 OS/Arch:      linux/amd64

$ sigil -v
0.4.0
```

**Comment by michaelshobbs** [comment]
Is `/tmp/user/1000` literally the directory it's trying to delete?

**Comment by michaelshobbs** [comment]
Also does the app deployment complete?

**Comment by michaelshobbs** [comment]
It would also be helpful to see the entirety of the `set -x` output of the `checks-deploy` trigger.

**Comment by seantanly** [comment]
> Is /tmp/user/1000 literally the directory it's trying to delete?

Yes I believe so, as per the output. I noticed in previous successful app startups/deployments the /tmp/user/1000 is created during the process and removed thereafter.

> Also does the app deployment complete?

Everything up to the point of cleanup seems to be working fine, the CHECKS completed successfully. Normally after the CHECKS are successful, the new container is made available. However, this in case whereby the cleanup failed, the new container seems to be killed as per the output.

> It would also be helpful to see the entirety of the set -x output of the checks-deploy trigger.

Here's the gist of entire Dokku output with trace on.
https://gist.github.com/seantanly/8e726b2532ae30c0423c15236f94de33

**Comment by michaelshobbs** [comment]
This output lines up with how we manage `TMPDIR` in the `cleanup()` call.
https://gist.github.com/seantanly/8e726b2532ae30c0423c15236f94de33#file-dokku_output-log-L1502-L1506

Something isn't lining up here and I am unable to reproduce this in any other environment. Can you please provide ALL of the information requested in our issue template? This will help provide more context around your specific installation.

**Comment by michaelshobbs** [comment]
Honestly it looks like something else is failing and the trap is getting triggered with no container id and another TMPDIR is defined. There's probably something we can do better to play nice here but I'll need more context to proceed.

**Comment by josegonzalez** [comment]
@seantanly can you also gist the following:

```
the plugin list
free -m
herokuish version
docker inspect
```

**Comment by michaelshobbs** [comment]
Issue template: https://github.com/dokku/dokku/blob/master/ISSUE_TEMPLATE.md

**Comment by seantanly** [comment]
Thanks guys for looking into this promptly. 

I have updated the gist with the info request under comments. Let me know you require more info.

**Comment by josegonzalez** [comment]
Is your app actually called `myapp`?

**Comment by seantanly** [comment]
No it is not, I have replaced the name.

**Comment by michaelshobbs** [comment]
I'm struggling to find where `/tmp/user/1000` is coming from....

**Comment by michaelshobbs** [comment]
We can definitely filter this output and error code.

```
rm -rf $TMPDIR &> /dev/null || true
```

@josegonzalez meh?

**Comment by seantanly** [comment]
Yea, it's weird, I think it's coming from `mktemp` which creates random file/directory names. I have also noticed `tmp/user/0` is always created as well. though the ownship differs `root:root` vs `tmp/user/1000` is `dokku:root`.

**Comment by michaelshobbs** [comment]
We don't use the pattern `tmp/user/...` anywhere

**Comment by seantanly** [comment]
I think that's by automated by `mktemp`

```
root@App0:~# mktemp -d -t hello.XXXXX
/tmp/user/0/hello.ctMwO
```

**Comment by michaelshobbs** [comment]
Whoa! What mktemp do you have?!

```
root@dokku:~/dokku# mktemp -d -t hello.XXXXX
/tmp/hello.dibh8
root@dokku:~/dokku# mktemp --version
mktemp (GNU coreutils) 8.21
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Jim Meyering and Eric Blake.
```

**Comment by michaelshobbs** [comment]
From `man mktemp`
```
-t     interpret TEMPLATE as a single file name component, relative to a directory: $TMPDIR, if set
```
Do you perhaps have `TMPDIR` set in the environment somewhere?

**Comment by michaelshobbs** [comment]
Also, @josegonzalez thinking we should probably rename `TMPDIR` wherever we use `mktemp`.

**Comment by josegonzalez** [comment]
@michaelshobbs that works for me. It seems like `cleanup` is being called twice? I'm looking at all @seantanly's other plugins - all official ones it seems -  to see how we might be triggering `cleanup` twice...

**Comment by josegonzalez** [comment]
Should we require the `id` as an argument to that `cleanup` command? And if not just bail (nicely)

**Comment by michaelshobbs** [comment]
@seantanly additionally I only see a call to `mktemp` once. Here: https://gist.github.com/seantanly/8e726b2532ae30c0423c15236f94de33#file-dokku_output-log-L1407

```
mktemp -d /tmp/dokku_CHECKS.XXXXX
```

**Comment by josegonzalez** [comment]
@seantanly Did you maybe scrub something from [these lines](https://gist.github.com/seantanly/8e726b2532ae30c0423c15236f94de33#file-dokku_output-log-L5188-L5193):

```
++ echo '=====> end myapp web container output'
=====> end myapp web container output
+ cleanup
+ declare 'desc=cleans up TMPDIR and print container output'
+ local id=
+ rm -rf /tmp/user/1000

**Comment by seantanly** [comment]
Since `mktemp -t` relies on $TMPDIR for the default directory, and I see that Dokku uses $TMPDIR as its own variable. Could there be a conflict?

@michaelshobbs Anyway, perhaps we should move away from -t since stated as [deprecated] on my `man mktemp`. We can do

```
mktemp --tmpdir -d hello.XXXXX
/tmp/user/0/hello.Ljlt
```
which lets the system use the default TMPDIR it desires, or you can set it with

```
root@App0:~# mktemp --tmpdir=/tmp -d hello.XXXXX
/tmp/hello.Xt5vD
```

Do note that `mktemp --tmpdir` expects the `tmpdir` to exist or it will fail.

**Comment by seantanly** [comment]
```
root@App0:~# mktemp --version
mktemp (GNU coreutils) 8.21
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Jim Meyering and Eric Blake.
```

**Comment by josegonzalez** [comment]
What is calling `mktemp -t`? We aren't. Please sign onto our slack room or IRC so we can give you quicker support.

**Comment by seantanly** [comment]
I just did a repository search and there's plenty in the code base, `mktemp -d -t`.

Going onto IRC

**Comment by josegonzalez** [comment]
Turns out we didn't reset the trap on successful exit, so it was being called twice because of the double returns?

**Comment by josegonzalez** [comment]
Fixed in 0.5.3
